openedx_learning/apps/authoring/containers/api.py

```diff
@@ -0,0 +1,334 @@
+            draft_version_pks=[None] * len(publishable_entities_pk),
+            published_version_pks=[None] * len(publishable_entities_pk),
+        )
+        container_version = ContainerEntityVersion.objects.create(
+            publishable_entity_version=publishable_entity_version,
+            container_id=container_pk,
+            defined_list=entity_list,
+            initial_list=entity_list,
+            frozen_list=None,
+        )
+    return container_version
+
+
+def create_next_container_version(
+    container_pk: int,
+    title: str,
+    publishable_entities_pk: list[int],
+    entity: PublishableEntity,
+    created: datetime,
+    created_by: int | None,
+) -> ContainerEntityVersion:
+    """
+    Create the next version of a container.
+
+    Args:
+        container_pk: The ID of the container to create the next version of.
+        title: The title of the container.
+        publishable_entities_pk: The IDs of the members of the container.
+        entity: The entity that the container belongs to.
+        created: The date and time the container version was created.
+        created_by: The ID of the user who created the container version.
+
+    Returns:
+        The newly created container version.
+    """
+    with atomic():
+        container = ContainerEntity.objects.get(pk=container_pk)
+        last_version = container.versioning.latest
+        next_version_num = last_version.version_num + 1
+        publishable_entity_version = publishing_api.create_publishable_entity_version(
+            entity.pk,
+            version_num=next_version_num,
+            title=title,
+            created=created,
+            created_by=created_by,
+        )
+        # This implementation assumes:
+        # 1. The changes provoking a new version are the addition, removal of members or reordering.
+        # 2. Published and draft versions are always the latest for all members.
+        # 3. When creating a new version, a new user-defined entity list is created to preserve the latest state as the previous user-defined list.
+        # TODO: instead consider copying the previous user-defined list as the frozen list, and add/remove to the previous user-defined list.
+        # If it's a reordering, the previous user-defined list is copied as the frozen and a new user-defined list is created with the new order.
+        new_user_defined_list = create_entity_list_with_rows(
+            entity_pks=publishable_entities_pk,
+            draft_version_pks=[None] * len(publishable_entities_pk),
+            published_version_pks=[None] * len(publishable_entities_pk),
+        )
+        container_version = ContainerEntityVersion.objects.create(
+            publishable_entity_version=publishable_entity_version,
+            container_id=container_pk,
+            defined_list=new_user_defined_list,
+            initial_list=last_version.initial_list,
+            frozen_list=last_version.defined_list,
+        )
```

Comment:
@ormsbee: this is a naive approach to the implementation for the sake of simplicity. Now, I'll be modifying it to comply with the requirements I've compiled in the cover letter. Although, I still have some questions about the approach:

_The defined_list is the EntityList that the author defines. This should never change, even if the things it references get soft-deleted (because we'll need to maintain it for reverts)._

**Questions** 
1. How do we ensure this doesn't change even if references get soft deleted?
2. Is the discard operation (made changes that created a new draft version) different from revert?

_initial_list is an EntityList where all the versions are pinned, to show what the exact versions of the children were at the time that the Container was created. We could technically derive this, but it would be awkward to query._ 

**Questions**
1. Let's say a version for a container is created referencing the latest versions for all members. Then, we manually pin all members when creating this list. Pinning a version would mean pointing the EntityListRow draft/published version to the latest versions at the time for this particular case, am I right? 

 _frozen_list is mutable if and only if there are unpinned references in defined_list. In that case, frozen_list should start as None, and be updated to pin references when another version of this Container becomes the Draft version. But if this version ever becomes the Draft *again* (e.g. the user hits "discard changes" or some kind of revert happens), then we need to clear this back to None._

**Questions**
1. Even if there's an unpinned reference, should frozen_list start as None?
2. When a new version is created and frozen_list is None, then we should pin all references from the defined_list when creating the new frozen_list?
3. Should we explicitly check when a version becomes draft **again**, so we can clear the frozen list? How could we detect a version becomes a draft again after a revert?